java2go.go

```diff
@@ -0,0 +1,209 @@
+package datetranslate
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// Translate from SimpleDateFormat [1] to ctimefmt [2].
+//
+// [1] https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html
+// [2] https://github.com/observIQ/ctimefmt/blob/4cb1bdfd4b74804fd68c5169d48a8db1e06cfe3e/ctimefmt.go#L63-L97
+func SimpleDateFormat2Ctimefmt(in string) (string, error) {
+	var sb strings.Builder
+
+	for len(in) > 0 {
+		s, err := getLongestCharacterStreak(in)
+		if err != nil {
+			return "", err
+		}
+
+		if s[0] == '\'' {
+			// Get the string from single quotes.
+			out, n, err := getFromQuotes(in)
+			if err != nil {
+				return "", fmt.Errorf("problem reading from quotes: %w", err)
+			}
+			if _, err = sb.WriteString(out); err != nil {
+				return "", fmt.Errorf("problem writing to string builder: %w", err)
+			}
+			in = in[n:]
+
+		} else if len(strings.TrimSpace(s)) == 0 {
+			// If we've reached the a streak of whitespace just copy it over.
+			if _, err = sb.WriteString(s); err != nil {
+				return "", fmt.Errorf("problem writing to string builder: %w", err)
+			}
+			in = in[len(s):]
+
+		} else {
+			translated, err := simpleDateFormat2CtimefmtSegment(s)
+			if err != nil {
+				return "", err
+			}
+			if _, err = sb.WriteString(translated); err != nil {
+				return "", fmt.Errorf("problem writing to string builder: %w", err)
+			}
+			in = in[len(s):]
+
+		}
+	}
+
+	return sb.String(), nil
+}
+
+// simpleDateFormat2CtimefmtSegment translates one segment into ctimefmt format.
+func simpleDateFormat2CtimefmtSegment(s string) (string, error) {
+	switch s {
+
+	// Years
+	case "y", "yyy", "yyyy", "Y", "YYY", "YYYY":
+		return "%Y", nil
+	case "yy", "YY":
+		return "%y", nil
+
+	// Months
+	case "M":
+		return "%q", nil
+	case "MM":
+		return "%m", nil
+	case "MMM":
+		return "%b", nil
+	case "MMMM":
```

Comment:
I'm just wondering, shouldn't the order here be `MMMM`, `MMM`, `MM`, `M`? (and similarly for other cases with the same character)

Reply:
Why would you be concerned about the order?

Reply:
I was wondering if e.g. "MMMM" would not be interpreted as 4x "M", since it's on the top of the list. But then this is not matching of a substring but a token returned by `getLongestCharacterStreak` so we should be good

Reply:
Ok it just got to me. You're right. I've changed it (although as you've noted it doesn't matter now).